File: backend/services/post-ts-koa-mongo-fdm/src/get-identity/impl.ts
import { Deps } from "../common/utils";
import { Workflow } from ".";
import { UserId, UserName } from "../common";
import * as db from "../common/utils/db";
import { WithId } from "mongodb";
import * as runtypes from "runtypes";
import { invalidAuth } from "../common/api/auth";

export class WorkflowImpl extends Workflow {
    private static readonly userHasName = runtypes.Record({
        name: runtypes.String,
    });
    async getUserName(id: UserId): Promise<UserName> {
        const oid = db.tryParseId(id);
        if (oid === undefined) {
            throw invalidAuth();
        }
        const user: WithId<unknown> | null = await this.deps.mongoDb.collection(db.users).findOne({ _id: oid });
        if (user === null) {
            throw invalidAuth();
        }
        db.validate(WorkflowImpl.userHasName, db.users, user);
        return user.name as UserName;
    }

    constructor(private deps: Deps) { super(); }
}

File: backend/services/post-ktor-mongo-fdm/src/main/kotlin/me/jason5lee/post_ktor_mongo_fdm/get_identity/impl.kt
package me.jason5lee.post_ktor_mongo_fdm.get_identity

import com.mongodb.client.model.Projections
import me.jason5lee.post_ktor_mongo_fdm.common.UserId
import me.jason5lee.post_ktor_mongo_fdm.common.UserName
import me.jason5lee.post_ktor_mongo_fdm.common.api.invalidAuth
import me.jason5lee.post_ktor_mongo_fdm.common.utils.Db
import me.jason5lee.post_ktor_mongo_fdm.common.utils.Deps

class WorkflowImpl(private val deps: Deps) : Workflow() {
    override suspend fun getUserName(id: UserId): UserName {
        val doc = Db.findById(
            db = deps.mongoDb,
            collection = Db.users,
            id = Db.toObjectIdOrNull(id.value) ?: throw invalidAuth(),
            projection = Projections.include("name")
        ) ?: throw invalidAuth()
        return UserName(doc.get("name"))
    }
}
